web/modules/custom/state_machine_revisions/src/EventSubscriber/WorkflowTransitionEventSubscriber.php

```diff
@@ -0,0 +1,125 @@
+<?php
+
+namespace Drupal\state_machine_revisions\EventSubscriber;
+
+use Drupal\Core\Entity\ContentEntityInterface;
+use Drupal\Core\Entity\EntityPublishedInterface;
+use Drupal\Core\Entity\RevisionableInterface;
+use Drupal\state_machine\Event\WorkflowTransitionEvent;
+use Drupal\state_machine\Plugin\Workflow\WorkflowInterface;
+use Drupal\state_machine\Plugin\Workflow\WorkflowState;
+use Drupal\state_machine_revisions\RevisionManagerInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Event subscriber to handle revisions on workflow-enabled entities.
+ */
+class WorkflowTransitionEventSubscriber implements EventSubscriberInterface {
+
+  /**
+   * The revision manager.
+   *
+   * @var \Drupal\state_machine_revisions\RevisionManagerInterface
+   */
+  protected $revisionManager;
+
+  /**
+   * Constructs a new EntityRevisionConverter object.
+   *
+   * @param \Drupal\state_machine_revisions\RevisionManagerInterface $revisionManager
+   *   The revision manager.
+   */
+  public function __construct(RevisionManagerInterface $revisionManager) {
+    $this->revisionManager = $revisionManager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents() {
+    return [
+      'state_machine.pre_transition' => 'handleRevision',
+    ];
+  }
+
+  /**
+   * Sets an entity revision to be the default based on the workflow.
+   *
+   * @param \Drupal\state_machine\Event\WorkflowTransitionEvent $event
+   *   The state change event.
+   */
+  public function handleRevision(WorkflowTransitionEvent $event) {
+    $entity = $event->getEntity();
+
+    // Bail out if this entity doesn't implement revisions. We do not limit
+    // strictly to content entities.
+    if (!$entity instanceof RevisionableInterface) {
+      return;
+    }
+
+    // Since all content entities implement the RevisionableInterface, we have
+    // to check if the entity has the revision support and if it is set to
+    // create a new revision.
+    if (!$entity->getEntityType()->isRevisionable() || !$entity->isNewRevision()) {
+      return;
+    }
+
+    // Verify if the new state is marked as published state.
+    $is_published_state = $this->isPublishedState($event->getToState(), $event->getWorkflow());
+    // Set revision as default when:
+    // - the entity is new;
+    // - the new state is a published one;
+    // - the current default revision is not published.
+    $entity->isDefaultRevision($entity->isNew() || $is_published_state || !$this->hasPublishedDefaultRevision($entity));
```

Comment:
Every time I see this method call it creeps me out. https://www.drupal.org/node/2706337